src/constructors/regular.c

```diff
@@ -134,6 +134,145 @@ igraph_error_t igraph_star(igraph_t *graph, igraph_integer_t n, igraph_star_mode
+ *           invalid number of vertices.
+ *         \cli IGRAPH_EINVAL
+ *           invalid center vertex.
+ *         \cli IGRAPH_EINVMODE
+ *           invalid mode argument.
+ *         \endclist
+ *
+ * Time complexity: O(|V|), the
+ * number of vertices in the graph.
+ *
+ * \sa \ref igraph_lattice(), \ref igraph_ring(), \ref igraph_star, \ref igraph_tree()
+ * for creating other regular structures.
+ *
+ */
+
+igraph_error_t igraph_wheel(igraph_t *graph, igraph_integer_t n, igraph_wheel_mode_t mode,
+                igraph_integer_t center) {
+
+    igraph_integer_t  v;
+    igraph_star_mode_t star_mode;
+    igraph_vector_int_t rim_edges = IGRAPH_VECTOR_NULL;
+    igraph_integer_t i;
+
+    /* create a star and also make use of its existing pre-check */
+    /* works for all input parameters                            */
+
+    switch(mode)
+    {
+        case IGRAPH_WHEEL_OUT:
+            star_mode = IGRAPH_STAR_OUT;
+            break;
+        case IGRAPH_WHEEL_IN:
+            star_mode = IGRAPH_STAR_IN;
+            break;
+        case IGRAPH_WHEEL_MUTUAL:
+            star_mode = IGRAPH_STAR_MUTUAL;
+            break;
+        case IGRAPH_WHEEL_UNDIRECTED:
+            star_mode = IGRAPH_STAR_UNDIRECTED;
+            break;
+        default:
+            IGRAPH_ERROR("invalid mode", IGRAPH_EINVMODE);
+    }
+
+    IGRAPH_CHECK(igraph_star(graph, n, star_mode, center));
+
+    /* If n <= 1, wheel graph is identical with star graph, */
+    /* no further processing is needed                      */
+
+    if (n <= 1) {
+        return IGRAPH_SUCCESS;
+    }
+
+    /* Add edges to the rim */
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&rim_edges, 2 * (n-1));
+
+    /* Assign first n-1 edges */
+
+    for (i = 0; i < n-2; i++) {
+        if ( i < center ) {
+            VECTOR(rim_edges)[2 * i] = i;
+            if ( i + 1 < center ) {
+                VECTOR(rim_edges)[2 * i + 1] = i + 1;
+            } else {
+                VECTOR(rim_edges)[2 * i + 1] = i + 2;
+            }
+        } else {
+            VECTOR(rim_edges)[2 * i] = i + 1;
+            VECTOR(rim_edges)[2 * i + 1] = i + 2;
+        }
+    }
+
+    /* Assign the last edge */
+
+    if ( n - 2 < center ) {
+        VECTOR(rim_edges)[2 * n - 4] = n - 2;
+    } else {
+        VECTOR(rim_edges)[2 * n - 4] = n - 1;
+    }
+    if ( center > 0 ) {
+        VECTOR(rim_edges)[2 * n - 3] = 0;
+    } else {
+        VECTOR(rim_edges)[2 * n - 3] = 1;
+    }
+
+    /* Combine the rim into the star to make it a wheel graph */
+
+    IGRAPH_CHECK(igraph_add_edges(graph, &rim_edges, 0));
+
+    /* Add a reverse direction rim if mode is MUTUAL */
+
+    if (mode == IGRAPH_WHEEL_MUTUAL) {
+        for (i=0; i < n-1; i++) {
+            v = VECTOR(rim_edges)[2 * i];
+            VECTOR(rim_edges)[2 * i] = VECTOR(rim_edges)[2 * i + 1];
+            VECTOR(rim_edges)[2 * i + 1] = v;
+        }
+        IGRAPH_CHECK(igraph_add_edges(graph, &rim_edges, 0));
+    }
```

Comment:
Since `igraph_add_edges` is quite expensive, can we use a single `igraph_add_edges` call instead of two? This can be achieved by using a larger `rim_edges` vector.

Reply:
Thanks for your advice. Rewrote and found that this approach is really better than previous approach.